### PR TITLE
Implement multi-stream recordings

### PR DESCRIPTION
**Pull request by emveepee**
Enable API changes for multiple recording stream . Allow optional Kodi extraction of thumbnails

### CODE REVIEW

File: src/buffers/RecordingBuffer.h [hunk 1]
```diff
@@ -74,11 +74,13 @@ namespace timeshift {
...
 
-    bool Open(const std::string inputUrl, const kodi::addon::PVRRecording& recording);
+    bool Open(const std::string inputUrl, const kodi::addon::PVRRecording& recording, int64_t streamId);
 
     std::atomic<bool> m_isLive;
 
     // recording start time
     time_t m_recordingTime;
+
+    int m_streamId;
```
Comment (phunkyfish) [inline src/buffers/RecordingBuffer.h]: Where is this variable used? Did you intend to store the passed in value?

Comment (emveepee) [reply src/buffers/RecordingBuffer.h]: It is not used,  I was going to use it to retrieve and delete the map for single streams but I just used the current stream instead.

Comment (phunkyfish) [reply src/buffers/RecordingBuffer.h]: Ok, then remove the changes to these files as they are not required.

Comment (emveepee) [reply src/buffers/RecordingBuffer.h]: Alright I removed that.

File: src/pvrclient-nextpvr.h [hunk 2]
```diff
@@ -76,15 +76,18 @@ class ATTR_DLL_LOCAL cPVRClientNextPVR : public kodi::addon::CInstancePVRClient
   int64_t LengthLiveStream() override;
   bool CanPauseStream() override;
   void PauseStream(bool paused) override;
+  PVR_ERROR  PauseRecordedStream(int64_t streamId, bool paused) override;
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.h]: 2 spaces after PVR_ERROR, should be one.

File: src/pvrclient-nextpvr.cpp [hunk 3]
```diff
@@ -720,12 +726,25 @@ void cPVRClientNextPVR::PauseStream(bool bPaused)
   if (IsServerStreaming())
   {
     if (m_nowPlaying == Recording)
-      m_recordingBuffer->PauseStream(bPaused);
+      m_multistreamRecording[m_streamCount]->PauseStream(bPaused);
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp:732]: `std::lock_guard<std::recursive_mutex> lock(m_multistreamRecording);` prior to accessing map.

No need to stop the lock, happens automatically at the end block

File: src/pvrclient-nextpvr.cpp [hunk 4]
```diff
@@ -720,12 +726,25 @@ void cPVRClientNextPVR::PauseStream(bool bPaused)
...
   }
 }
 
+
+PVR_ERROR cPVRClientNextPVR::PauseRecordedStream(int64_t streamId, bool bPaused)
+{
+  if (IsServerStreaming())
+  {
+    if (m_nowPlaying == Recording)
+      m_multistreamRecording[streamId]->PauseStream(bPaused);
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp:747]: `std::lock_guard<std::recursive_mutex> lock(m_multistreamRecording);` prior to accessing map.

No need to stop the lock, happens automatically at the end block

File: src/pvrclient-nextpvr.cpp [hunk 5]
```diff
@@ -743,45 +762,58 @@ bool cPVRClientNextPVR::CanSeekStream(void)
 bool cPVRClientNextPVR::OpenRecordedStream(const kodi::addon::PVRRecording& recording, int64_t& streamId)
 {
   kodi::addon::PVRRecording copyRecording = recording;
-  m_nowPlaying = Recording;
   copyRecording.SetDirectory(m_recordings.m_hostFilenames[recording.GetRecordingId()]);
   const std::string line = kodi::tools::StringUtils::Format("%s/live?recording=%s&client=XBMC-%s", m_settings->m_urlBase, recording.GetRecordingId().c_str(), m_request.GetSID());
-  return m_recordingBuffer->Open(line, copyRecording);
+  m_mutexMulti.lock();
+  m_nowPlaying = Recording;
+  m_multistreamRecording.emplace(++m_streamCount, new timeshift::RecordingBuffer(m_settings, m_request));
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp:776]: `std::lock_guard<std::recursive_mutex> lock(m_multistreamRecording);` prior to accessing map.

No need to stop the lock, happens automatically at the end block

File: src/pvrclient-nextpvr.cpp [hunk 6]
```diff
@@ -743,45 +762,58 @@ bool cPVRClientNextPVR::CanSeekStream(void)
...
 
 void cPVRClientNextPVR::CloseRecordedStream(int64_t streamId)
 {
-  if (IsServerStreamingRecording())
+  if (IsServerStreamingRecording(streamId))
   {
-    m_recordingBuffer->Close();
-    m_recordingBuffer->SetDuration(0);
+    m_mutexMulti.lock();
+    m_multistreamRecording[streamId]->Close();
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp:791]: `std::lock_guard<std::recursive_mutex> lock(m_multistreamRecording);` prior to accessing map.

No need to stop the lock, happens automatically at the end block

File: src/pvrclient-nextpvr.cpp [hunk 7]
```diff
@@ -743,45 +762,58 @@ bool cPVRClientNextPVR::CanSeekStream(void)
...
+    m_multistreamRecording.erase(streamId);
+    m_mutexMulti.unlock();
   }
-  m_nowPlaying = NotPlaying;
+  if (m_multistreamRecording.size() == 0)
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp:794]: `std::lock_guard<std::recursive_mutex> lock(m_multistreamRecording);` prior to accessing map.

No need to stop the lock, happens automatically at the end block

File: src/pvrclient-nextpvr.cpp [hunk 8]
```diff
@@ -743,45 +762,58 @@ bool cPVRClientNextPVR::CanSeekStream(void)
...
+  kodi::Log(ADDON_LOG_DEBUG, "Closed streamId %d remaining %d", streamId, m_multistreamRecording.size());
 }
 
 int cPVRClientNextPVR::ReadRecordedStream(int64_t streamId, unsigned char* pBuffer, unsigned int iBufferSize)
 {
-  if (IsServerStreamingRecording())
+  if (IsServerStreamingRecording(streamId))
   {
-    return m_recordingBuffer->Read(pBuffer, iBufferSize);
+    return m_multistreamRecording[streamId]->Read(pBuffer, iBufferSize);
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp:804]: `std::lock_guard<std::recursive_mutex> lock(m_multistreamRecording);` prior to accessing map.

No need to stop the lock, happens automatically at the end block

File: src/pvrclient-nextpvr.cpp [hunk 9]
```diff
@@ -743,45 +762,58 @@ bool cPVRClientNextPVR::CanSeekStream(void)
...
   return -1;
 }
 
 int64_t cPVRClientNextPVR::SeekRecordedStream(int64_t streamId, int64_t iPosition, int iWhence)
 {
-  if (IsServerStreamingRecording())
+  if (IsServerStreamingRecording(streamId))
   {
-    return m_recordingBuffer->Seek(iPosition, iWhence);
+    return m_multistreamRecording[streamId]->Seek(iPosition, iWhence);
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp:814]: `std::lock_guard<std::recursive_mutex> lock(m_multistreamRecording);` prior to accessing map.

No need to stop the lock, happens automatically at the end block

File: src/pvrclient-nextpvr.cpp [hunk 10]
```diff
@@ -800,19 +832,45 @@ bool cPVRClientNextPVR::IsRealTimeStream()
   if (IsServerStreaming())
   {
     if (m_nowPlaying == Recording)
-      return m_recordingBuffer->IsRealTimeStream();
+      return m_multistreamRecording[m_streamCount]->IsRealTimeStream();
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp:844]: `std::lock_guard<std::recursive_mutex> lock(m_multistreamRecording);` prior to accessing map.

No need to stop the lock, happens automatically at the end block

File: src/pvrclient-nextpvr.cpp [hunk 11]
```diff
@@ -800,19 +832,45 @@ bool cPVRClientNextPVR::IsRealTimeStream()
...
   return false;
 }
-
 PVR_ERROR cPVRClientNextPVR::GetStreamTimes(kodi::addon::PVRStreamTimes& stimes)
 {
   if (IsServerStreaming())
   {
     if (m_nowPlaying == Recording)
-      return m_recordingBuffer->GetStreamTimes(stimes);
+      return m_multistreamRecording[m_streamCount]->GetStreamTimes(stimes);
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp:858]: `std::lock_guard<std::recursive_mutex> lock(m_multistreamRecording);` prior to accessing map.

No need to stop the lock, happens automatically at the end block

File: src/pvrclient-nextpvr.cpp [hunk 12]
```diff
@@ -800,19 +832,45 @@ bool cPVRClientNextPVR::IsRealTimeStream()
...
+}
+
+
+PVR_ERROR cPVRClientNextPVR::IsRecordedStreamRealTime(int64_t streamId, bool& isRealTime)
+{
+  if (IsServerStreaming())
+  {
+    if (m_nowPlaying == Recording)
+    {
+      isRealTime = m_multistreamRecording[streamId]->IsRealTimeStream();
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp:874]: `std::lock_guard<std::recursive_mutex> lock(m_multistreamRecording);` prior to accessing map.

No need to stop the lock, happens automatically at the end block

File: src/pvrclient-nextpvr.cpp [hunk 13]
```diff
@@ -800,19 +832,45 @@ bool cPVRClientNextPVR::IsRealTimeStream()
...
+  }
+  return PVR_ERROR_NO_ERROR;
+}
+
+PVR_ERROR cPVRClientNextPVR::GetRecordedStreamTimes(int64_t streamId, kodi::addon::PVRStreamTimes& stimes)
+{
+  if (IsServerStreaming())
+  {
+    if (m_nowPlaying == Recording)
+      return m_multistreamRecording[streamId]->GetStreamTimes(stimes);
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp:889]: `std::lock_guard<std::recursive_mutex> lock(m_multistreamRecording);` prior to accessing map.

No need to stop the lock, happens automatically at the end block

File: src/pvrclient-nextpvr.cpp [hunk 14]
```diff
@@ -836,11 +894,11 @@ PVR_ERROR cPVRClientNextPVR::GetStreamReadChunkSize(int& chunksize)
 
 bool cPVRClientNextPVR::IsServerStreaming()
 {
-  if (IsServerStreamingLive(false) || IsServerStreamingRecording(false))
+  if (IsServerStreamingLive(false) || m_multistreamRecording.size() > 0)
   {
     return true;
   }
-  kodi::Log(ADDON_LOG_ERROR, "Unknown streaming state %d %d %d", m_nowPlaying, m_recordingBuffer->GetDuration(), !m_livePlayer);
+  kodi::Log(ADDON_LOG_ERROR, "Unknown streaming state %d %d %d", m_nowPlaying, m_multistreamRecording.size(), !m_livePlayer);
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp:918]: `std::lock_guard<std::recursive_mutex> lock(m_multistreamRecording);` prior to accessing map.

No need to stop the lock, happens automatically at the end block

File: src/pvrclient-nextpvr.cpp [hunk 15]
```diff
@@ -851,29 +909,21 @@ bool cPVRClientNextPVR::IsServerStreamingLive(bool log)
     return true;
   }
   if (log)
-    kodi::Log(ADDON_LOG_ERROR, "Unknown live streaming state %d %d %d", m_nowPlaying, m_recordingBuffer->GetDuration(), !m_livePlayer);
+    kodi::Log(ADDON_LOG_ERROR, "Unknown live streaming state %d %d %d", m_nowPlaying, m_multistreamRecording.size(), !m_livePlayer);
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp:929]: `std::lock_guard<std::recursive_mutex> lock(m_multistreamRecording);` prior to accessing map.

No need to stop the lock, happens automatically at the end block

File: src/pvrclient-nextpvr.cpp [hunk 16]
```diff
@@ -851,29 +909,21 @@ bool cPVRClientNextPVR::IsServerStreamingLive(bool log)
...
   return false;
 }
 
-bool cPVRClientNextPVR::IsServerStreamingRecording(bool log)
+bool cPVRClientNextPVR::IsServerStreamingRecording(int64_t streamId, bool log)
 {
-  if (m_nowPlaying == Recording && m_recordingBuffer->GetDuration() > 0)
+  if (m_nowPlaying == Recording && m_multistreamRecording.size() > 0)
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp:935]: `std::lock_guard<std::recursive_mutex> lock(m_multistreamRecording);` prior to accessing map.

No need to stop the lock, happens automatically at the end block

File: src/pvrclient-nextpvr.cpp [hunk 17]
```diff
@@ -851,29 +909,21 @@ bool cPVRClientNextPVR::IsServerStreamingLive(bool log)
...
   {
-    return true;
+    return m_multistreamRecording.find(streamId) != m_multistreamRecording.end();
   }
   if (log)
-    kodi::Log(ADDON_LOG_ERROR, "Unknown recording streaming state %d %d %d", m_nowPlaying, m_recordingBuffer->GetDuration(), !m_livePlayer);
+    kodi::Log(ADDON_LOG_ERROR, "Unknown recording streaming state %d %d %d", m_nowPlaying, m_multistreamRecording.size(), !m_livePlayer);
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp:941]: `std::lock_guard<std::recursive_mutex> lock(m_multistreamRecording);` prior to accessing map.

No need to stop the lock, happens automatically at the end block

File: src/pvrclient-nextpvr.cpp [hunk 18]
```diff
@@ -743,45 +762,58 @@ bool cPVRClientNextPVR::CanSeekStream(void)
...
+  m_mutexMulti.lock();
+  m_nowPlaying = Recording;
+  m_multistreamRecording.emplace(++m_streamCount, new timeshift::RecordingBuffer(m_settings, m_request));
+  streamId = m_streamCount;
+  bool ret = m_multistreamRecording[streamId]->Open(line, copyRecording, streamId);
+  if (!ret)
+  {
+    CloseRecordedStream(streamId);
+  }
+  m_mutexMulti.unlock();
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp]: remove this line. Not required as lock will automatically be freed.

File: src/pvrclient-nextpvr.cpp [hunk 19]
```diff
@@ -743,45 +762,58 @@ bool cPVRClientNextPVR::CanSeekStream(void)
...
 {
-  if (IsServerStreamingRecording())
+  if (IsServerStreamingRecording(streamId))
   {
-    m_recordingBuffer->Close();
-    m_recordingBuffer->SetDuration(0);
+    m_mutexMulti.lock();
+    m_multistreamRecording[streamId]->Close();
+    m_multistreamRecording.erase(streamId);
+    m_mutexMulti.unlock();
```
Comment (phunkyfish) [inline src/pvrclient-nextpvr.cpp]: remove this line. Not required as lock will automatically be freed.

File: src/buffers/RecordingBuffer.h (hunk 1 above)
Comment (phunkyfish) [reply src/buffers/RecordingBuffer.h]: Still not removed. Likely don't need changes in either `RecordingBuffer` file.

**Review by phunkyfish — CHANGES_REQUESTED** [review]

**Review by phunkyfish — APPROVED** [review]

**Pull request merged**